src/main/java/frc/robot/subsystems/TurretSubsystem.java

```diff
@@ -0,0 +1,246 @@
+// Copyright (c) FIRST and other WPILib contributors.
+// Open Source Software; you can modify and/or share it under the terms of
+// the WPILib BSD license file in the root directory of this project.
+
+package frc.robot.subsystems;
+
+import com.ctre.phoenix6.hardware.CANcoder;
+import com.revrobotics.CANSparkBase.SoftLimitDirection;
+import com.revrobotics.CANSparkLowLevel.MotorType;
+import com.revrobotics.CANSparkMax;
+import edu.wpi.first.hal.SimDouble;
+import edu.wpi.first.math.MathUtil;
+import edu.wpi.first.math.controller.PIDController;
+import edu.wpi.first.math.geometry.Pose2d;
+import edu.wpi.first.math.system.plant.DCMotor;
+import edu.wpi.first.math.util.Units;
+import edu.wpi.first.networktables.GenericEntry;
+import edu.wpi.first.wpilibj.RobotController;
+import edu.wpi.first.wpilibj.shuffleboard.Shuffleboard;
+import edu.wpi.first.wpilibj.shuffleboard.ShuffleboardTab;
+import edu.wpi.first.wpilibj.simulation.SimDeviceSim;
+import edu.wpi.first.wpilibj.simulation.SingleJointedArmSim;
+import edu.wpi.first.wpilibj.smartdashboard.Mechanism2d;
+import edu.wpi.first.wpilibj.smartdashboard.MechanismLigament2d;
+import edu.wpi.first.wpilibj.smartdashboard.MechanismRoot2d;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import edu.wpi.first.wpilibj.util.Color;
+import edu.wpi.first.wpilibj.util.Color8Bit;
+import edu.wpi.first.wpilibj2.command.Command;
+import edu.wpi.first.wpilibj2.command.InstantCommand;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.Constants;
+import frc.robot.Robot;
+import frc.robot.stateManagement.AllianceColor;
+import frc.robot.stateManagement.RobotStateManager;
+import frc.robot.utilities.DebugEntry;
+import frc.robot.utilities.LimelightHelpers;
+import java.util.function.Consumer;
+
+public class TurretSubsystem extends SubsystemBase {
+
+  private CANSparkMax turretMotor;
+  private SingleJointedArmSim turretSim;
+  private Mechanism2d turretMech;
+  private MechanismRoot2d root;
+  private MechanismLigament2d turretAngleSim;
+  private ShuffleboardTab turretTab = Shuffleboard.getTab("Turret Tab");
+  private SimDeviceSim simEncoder;
+  private SimDouble simTurretPos;
+
+  private PIDController turretPIDController;
+  private CANcoder m_encoder;
+  private double turretPosition;
+  private double turretVelocity;
+  private Consumer<Double> testPosition;
+  private ShuffleboardTab configTab = Shuffleboard.getTab("Config");
+  private GenericEntry kP = configTab.add("Turret KP", Constants.TurretConstants.KP).getEntry();
+  private GenericEntry kI = configTab.add("Turret KI", Constants.TurretConstants.KI).getEntry();
+  private GenericEntry kD = configTab.add("Turret KD", Constants.TurretConstants.KD).getEntry();
```

Comment:
Optional suggestion:

Since PIDController implements Sendable, you can actually just put
`configTab.add("Turret PID", turretPIDController);` after you instantiate turretPIDController and that will add a widget entry that updates the PID values from shuffleboard auto-magically.